client/post-editor/editor-confirmation-sidebar/index.jsx

```diff
@@ -41,6 +41,8 @@ class EditorConfirmationSidebar extends Component {
 		status: PropTypes.string,
 	};
 
+	mountSidebar = false;
```

Comment:
I'm curious here why we're setting an instance variable on the component instead of working through `state`

Reply:
That was discussed in a related PR in https://github.com/Automattic/wp-calypso/pull/26227#discussion_r204676663.

Maintaining the `mountSidebar` flag in state requires more code that's not trivial to write, and instance property seems to do the job equally well.

Reply:
Thanks for the link @jsnajdr! If we're going to do this I think it would be pretty essential that we provide a good explanation in a comment why this is here as an instance variable.

In the link I was having a hard time following why this was problematic or why having it in local state would complicate things. Is this the chain we're wanting to avoid?

 1. Click callback on accordion to open it
 2. Callback calls `this.setState()`
 3. Some code runs before `setState()` executes?

I'm still curious if we aren't missing something.

```js
state = {
  hasExpanded: false,
  isExpanded: false,
}

getDerivedStateFromProps( props, state ) {
  if ( state.isExpanded || props.forceExpanded ) {
    return { hasExpanded: true };
  } 
}

render() {
  return hasExpanded && <Components />
}
```

is the added complexity just the inclusion of `getDerivedStateFromProps()`? If that's the case I know that I'd prefer having the logic in there where it won't stand out unexpectedly. Am I missing something big here? What synchronization issues do we have with props?

Reply:
Thanks for the suggestion, @dmsnell!

`getDerivedStateFromProps` is a much cleaner approach than what I had in mind, but I'd still argue against it, since deriving state is [strongly discouraged](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html) and best reserved to very specific use-cases.

It also seems like a large amount of effort for the sake of keeping that boolean in state, with what looks to me like a small set of benefits – mostly around ease of debugging, from what I can tell.

Conceptually, I'm not sure the boolean should be in state either. I see entries in state as being _things that should trigger a render_ when they change. This boolean seems to me more like a _thing that changes during a render_, i.e. merely a side-effect of how we do rendering. It doesn't change outside the context of an existing render, and thus should never trigger one.

Still, these are all personal opinions and preferences :) I'm more than happy to make the change if folks agree `getDerivedStateFromProps` is the best approach.

Reply:
I feel pretty strongly on this one since using instance variables is a good way to introduce unexpected behaviors and will definitely confuse people who stumble upon it - probably someone will break this component by not realizing what's happening; I'll cede this to @jsnajdr or @blowery or @ockham or someone else though.

> It also seems like a large amount of effort for the sake of keeping that boolean in state

It's a few additional lines to the instance variable and both cases run the same logic on every render. Is it really that different?

> deriving state is strongly discouraged

but…

> [it] exists for only one purpose. It enables a component to update its internal state as the result of changes in props

that's what we're doing here. we could do it in `componentDidUpdate()` which would trigger one additional render the first time the component mounts, but not that it will never be set more than once.

> I see entries in state as being things that should trigger a render when they change. This boolean seems to me more like a thing that changes during a render

nothing _should_ change during a render, nothing really _can_ change during a render because we build the virtual DOM before we commit it to the actual DOM. state changes just like prop changes should trigger a re-render if they effect the resulting visuals, which I'd argue this does.

note too that the only reason the instance variable works is because its only transition occurs on prop or state changes. I'd suggest that we're creating parallel state here whereas React has designed an idiomatic way to update state based on those changes.

Reply:
> In the link I was having a hard time following why this was problematic or why having it in local state would complicate things.

Hi Dennis, it's just that writing correct React lifecycle code is hard and there are many opportunities for subtle bugs. If you want to submit a PR to refactor the Accordion component, I'll be happy to review it and check for the edge cases.

Just be aware that you'll be refactoring code that currently gives correct results 100% of the time, although it's a bit unorthodox and some folks might have strong feelings about it. Be careful to not break it.

For example, with your implementation of `getDerivedStateFromProps`, you'll also need to modify the code that calls `setState` to toggle the expansion state. The existing version will not work as expected.

Reply:
> you'll also need to modify the code that calls setState to toggle the expansion state. The existing version will not work as expected.

Thanks for pointing this out - I'm not sure why it would break - can you help me understand?

Also, I thought that this PR was intentionally changing behavior, meaning that this change entirely is `refactoring code that currently gives correct results 100% of the time`

Reply:
> Thanks for pointing this out - I'm not sure why it would break - can you help me understand?

What if someone tried to figure it out and reported their findings? I created an issue (#27340) for the GM Calypso class crowd. There's a lot of learning opportunities there!

When suggesting to make a PR, I was referring to the Accordion component, where the instance-variable sin has already been committed some time ago (#26227).

Reply:
I rewrote the optimisation using the same approach as in #27555. Thanks, @dmsnell and @jsnajdr!